halo2/src/prover.rs

```diff
@@ -134,60 +194,119 @@ impl<'a, F: FieldElement> Halo2Prover<'a, F> {
 
         log::info!("Proof generation done.");
 
-        Ok(proof)
+        Ok((proof, publics))
+    }
+
+    /// Generate a single proof for a given PIL using Poseidon transcripts.
+    /// One or more of these proofs can be aggregated by `prove_snark_aggr`.
+    pub fn prove_poseidon(
+        &self,
+        witness: &[(String, Vec<F>)],
+        witgen_callback: WitgenCallback<F>,
+    ) -> Result<Vec<u8>, String> {
+        assert!(matches!(self.proof_type, ProofType::Poseidon));
+
+        let proof = self.prove::<_, aggregation::PoseidonTranscript<NativeLoader, _>,  aggregation::PoseidonTranscript<NativeLoader, _>>(witness, witgen_callback)?;
+        Ok(proof.0)
     }
 
-    pub fn prove_aggr(
+    /// Generate a single proof for a given PIL using Keccak transcripts.
+    /// These proofs can be verified directly on Ethereum.
+    pub fn prove_snark_single(
         &self,
         witness: &[(String, Vec<F>)],
+        witgen_callback: WitgenCallback<F>,
+    ) -> Result<Vec<u8>, String> {
+        assert!(matches!(self.proof_type, ProofType::SnarkSingle));
+
+        let proof = self.prove::<_, EvmTranscript<_, _, _, _>, EvmTranscript<G1Affine, _, _, _>>(
+            witness,
+            witgen_callback,
+        )?;
+
+        log::info!("Verifying SNARK in the EVM...");
+
+        let verifier_solidity = self.ethereum_verifier_single_snark()?;
+
+        let verifier_creation_code = compile_solidity(verifier_solidity);
+
+        let mut evm = Evm::default();
+        let verifier_address = evm.create(verifier_creation_code);
+
+        let calldata = encode_calldata(None, &proof.0, &proof.1[0]);
+
+        let (_gas_cost, output) = evm.call(verifier_address, calldata);
+        assert_eq!(output, [vec![0; 31], vec![1]].concat());
+
+        log::info!("EVM verification done.");
+
+        Ok(proof.0)
+    }
+
+    /// Generate a recursive proof that compresses one or more Poseidon proofs.
+    /// These proofs can be verified directly on Ethereum.
+    pub fn prove_snark_aggr(
+        &self,
+        witness: &[(String, Vec<F>)],
+        witgen_callback: WitgenCallback<F>,
         proof: Vec<u8>,
     ) -> Result<Vec<u8>, String> {
+        assert!(matches!(self.proof_type, ProofType::SnarkAggr));
+
         log::info!("Starting proof aggregation...");
 
         log::info!("Generating circuit for app snark...");
 
-        let circuit_app = PowdrCircuit::new(self.analyzed, self.fixed).with_witness(witness);
-        let publics = vec![circuit_app.instance_column()];
+        let circuit_app = PowdrCircuit::new(self.analyzed, self.fixed)
+            .with_witgen_callback(witgen_callback)
+            .with_witness(witness);
 
-        assert_eq!(publics.len(), 1);
-        if !publics[0].is_empty() {
+        // TODO Support publics in the app snark
```

Comment:
```suggestion
```

I think I added it, but it's kind of redundant with the `unimplemented!` call just below ;)

Reply:
Yea exactly that's what I meant here